### PR TITLE
Optimal choice of `apply_unitary` versus `apply_unitary_einsum`

### PR DESCRIPTION
**Pull request by albi3ro**
Upon investigating the differences between `dev.apply_unitary` and `dev.apply_unitary_einsum` on `"default.qubit"`, I believe the threshold for choosing between the two should be changed.  For lower device sizes, `einsum` is faster for all gate sizes.  Only above 8 device wires does `tensordot` become more efficient for larger gate sizes.

This profiling was done with a matrix generated by `scipy.stats.unitary_group.rvs(2**ii, random_state=5043)`. Seed does marginally affect the timing, but it does not switch the qualitative behaviour particularly much.

"Low" indicates gate wires `(0, 1, .., (gate_wires-1)`, whereas "high" indicates numbering `(device_wires-1, ..., device_wires-1-gate_wires)`.  I wanted to control for wire ordering, as lower dimensions may be easier to sum over.

image

image

image

I'll will continue this comparison for PyTorch, JAX, and TensorFlow.

### CODE REVIEW

**Comment by josh146** [comment]
Interesting @albi3ro 🤔 

Remember to take into account the interface and diff method if you haven't already. I would expect that `diff_method="parameter-shift"` and `diff_method="backprop"` will result in two different profiles *just on the forward pass*.

**Comment by albi3ro** [comment]
I was comparing 'dev._apply_unitary' to 'dev._apply_unitary_einsum' on '"default.qubit"'.

I was put on this track from profiling the tensorflow interface, so I'm investigating the interface devices next.

**Comment by josh146** [comment]
Hey @albi3ro, just double checking the status of this PR?

**Comment by albi3ro** [comment]
I got bogged down in exploring the different interfaces, then moved on to high priorities.  I'll try to revisit this soon since it'll probably help make things more efficient.
